pkg/mcp/sink/sink.go

```diff
@@ -0,0 +1,293 @@
+	// on the value of Incremental.
+	//
+	// When Incremental=True, the list only includes new/updated resources.
+	//
+	// When Incremental=False, the list includes the full list of resources.
+	// Any previously received resources not in this list should be deleted.
+	Objects []*Object
+
+	// List of deleted resources by name. The resource name corresponds to the
+	// resource's metadata name.
+	//
+	// Ignore when Incremental=false.
+	Removed []string
+
+	// When true, the set of changes represents an incremental resource update. The
+	// `Objects` is a list of added/update resources and `Removed` is a list of delete
+	// resources.
+	//
+	// When false, the set of changes represents a full-state update for the specified
+	// type. Any previous resources not included in this update should be removed.
+	Incremental bool
+}
+
+// Updater provides configuration changes in batches of the same protobuf message type.
+type Updater interface {
+	// Apply is invoked when the node receives new configuration updates
+	// from the server. The caller should return an error if any of the provided
+	// configuration resources are invalid or cannot be applied. The node will
+	// propagate errors back to the server accordingly.
+	Apply(*Change) error
+}
+
+// Stream is a stream interface for sending RequestResources messages and receiving Resource messages.
+type Stream interface {
+	Send(*mcp.RequestResources) error
+	Recv() (*mcp.Resources, error)
+}
+
+type perTypeState struct {
+	versions map[string]string
+	updated  bool
+}
+
+// Sink implements the resource sink message exchange for MCP. It can be instantiated by client and server
+// sink implementations to manage the MCP message exchange.
+type Sink struct {
+	mu    sync.Mutex
+	state map[string]*perTypeState
+
+	nodeInfo *mcp.SinkNode
+	updater  Updater
+	journal  recentRequestsJournal
+	metadata map[string]string
+	reporter monitoring.Reporter
+}
+
+// NewSink creates a new resource sink.
+func NewSink(supportedTypeURLs []string, updater Updater, id string, metadata map[string]string, reporter monitoring.Reporter) *Sink { // nolint: lll
+	nodeInfo := &mcp.SinkNode{
+		Id: id,
+		Metadata: &types.Struct{
+			Fields: map[string]*types.Value{},
+		},
+	}
+	for k, v := range metadata {
+		nodeInfo.Metadata.Fields[k] = &types.Value{
+			Kind: &types.Value_StringValue{StringValue: v},
+		}
+	}
+
+	state := make(map[string]*perTypeState)
+	for _, typeURL := range supportedTypeURLs {
+		state[typeURL] = &perTypeState{
+			versions: make(map[string]string),
+		}
+	}
+
+	return &Sink{
+		state:    state,
+		nodeInfo: nodeInfo,
+		updater:  updater,
+		metadata: metadata,
+		reporter: reporter,
+	}
+}
+
+// ResetResourceVersions resets the cached resource versions for all supported types. This function may be used
+// to effectively force full state delivery when a stream is reestablished.
+func (sink *Sink) ResetResourceVersions() {
+	sink.mu.Lock()
```

Comment:
When is a sink used in a multithread environment?

Reply:
The lock here protects the per-type resource version state. This state is used in the request path for fast-resume on re-connection. The caller can invoke `ResetResourceVersions()` to reset this version information and force a full state delivery. This function is invoked outside of the normal request path and must therefore protect access to the state with a lock.

Reply:
Given this is a utility package that several components will use, it'd worth stating the multithreading expectations/support explicitly in a comment.